File: 2 Add Two Numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2, bool carry = false) {
        //改用遞迴
        if(!l1 && !l2){
            if(carry)
                return new ListNode(1);
            return NULL;
        }
        if(!l2){
            swap(l1, l2);
        }
        if(!l1){
            l2->val += carry;
            carry = l2->val > 9;
            if(carry){
                l2->val -= 10;
                l2->next = addTwoNumbers(l1, l2->next, carry);
            }
            return l2;
        }
        else{
            l2->val += l1->val + carry;
            carry = l2->val > 9;
            if(carry)
                l2->val -= 10;
            l2->next = addTwoNumbers(l1->next, l2->next, carry);
            return l2;
        }
    }
};

File: 2 Add Two Numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1, l2, carry = False):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        if not l1 and not l2:
            if carry:
                return ListNode(1)
            return None
        if not l1:
            l1, l2 = l2, l1
        if not l2:
            l1.val += int(carry)
            carry = l1.val > 9;
            if carry:
                l1.val -= 10
                l1.next = self.addTwoNumbers(l1.next, None, carry)
            return l1
        else:
            l1.val += l2.val + int(carry)
            carry = l1.val > 9
            if carry:
                l1.val -= 10;
            l1.next = self.addTwoNumbers(l1.next, l2.next, carry)
            return l1